karapace/auth.py

```diff
@@ -0,0 +1,210 @@
+                    users,
+                )
+                self.permissions = permissions
+                log.info(
+                    "Loaded schema registry access control rules: %s",
+                    [(entry.username, entry.operation.value, entry.resource.pattern) for entry in permissions],
+                )
+        except Exception as ex:
+            raise InvalidConfiguration("Auth configuration is not valid") from ex
+
+    def check_authorization(self, user: Optional[User], operation: Operation, resource: str) -> bool:
+        if user is None:
+            return False
+
+        def check_operation(operation: Operation, aclentry: ACLEntry) -> bool:
+            if operation == Operation.Read:
+                return True
+            if aclentry.operation == Operation.Write:
+                return True
+            return False
+
+        def check_resource(resource: str, aclentry: ACLEntry) -> bool:
+            return aclentry.resource.match(resource) is not None
+
+        for aclentry in self.permissions:
+            if (
+                aclentry.username == user.username
+                and check_operation(operation, aclentry)
+                and check_resource(resource, aclentry)
+            ):
+                return True
+        return False
+
+    def authenticate(self, request: aiohttp.web.Request) -> User:
+        auth_header = request.headers.get("Authorization")
+        if auth_header is None:
+            raise aiohttp.web.HTTPUnauthorized(
+                headers={"WWW-Authenticate": 'Basic realm="Karapace Schema Registry"'},
+                text='{"message": "Unauthorized"}',
+                content_type=JSON_CONTENT_TYPE,
+            )
+        try:
+            auth = aiohttp.BasicAuth.decode(auth_header)
+        except ValueError:
+            # pylint: disable=raise-missing-from
+            raise aiohttp.web.HTTPUnauthorized(
+                headers={"WWW-Authenticate": 'Basic realm="Karapace Schema Registry"'},
+                text='{"message": "Unauthorized"}',
+                content_type=JSON_CONTENT_TYPE,
+            )
+        user = self.userdb.get(auth.login)
+        if user is None or not user.compare_password(auth.password):
+            raise aiohttp.web.HTTPUnauthorized(
+                headers={"WWW-Authenticate": 'Basic realm="Karapace Schema Registry"'},
+                text='{"message": "Unauthorized"}',
+                content_type=JSON_CONTENT_TYPE,
+            )
+
+        return user
+
+
+def main() -> int:
+    parser = argparse.ArgumentParser(prog="karapace_mkpasswd", description="Karapace password hasher")
+    parser.add_argument("-u", "--user", help="Username", type=str)
+    parser.add_argument(
+        "-a", "--algorithm", help="Hash algorithm", choices=["sha1", "sha256", "sha512", "scrypt"], default="sha512"
+    )
+    parser.add_argument(metavar="password", dest="plaintext_password", help="Password to hash", type=str)
```

Comment:
Just a note:
Probably fine for now, but later we need to ask a password interactively with `getpass` if this parameter is omitted, for better security.